File: 2020/day5.js
const fs = require('fs');

fs.readFile('day5-input', 'utf8', (_, data) => {
  const seats = data.trim().split('\n').map(seatID).sort((a, b) => b - a);

  console.log('Part 1:', seats[0]);

  for (let i = seats.length - 1; i >= 0; i--) {
    if (seats[i - 1] !== seats[i] + 1) {
      console.log('Part 2:', seats[i] + 1);
      break;
    }
  }
});

const seatID = str => getRow(str) * 8 + getCol(str);
const getRow = str => binarySpacePos('B', str.slice(0, 7));
const getCol = str => binarySpacePos('R', str.slice(7));

const binarySpacePos = (upperChar, str) => {
  const chars = str.split('');
  const len = chars.length;

  return chars.reduce((pos, char, index) => (
    char == upperChar ? pos + 2 ** (len - index - 1) : pos
  ), 0);
};

File: 2020/day5.hs
import Data.List (sort)

main :: IO ()
main = do
  let seatID str = row str * 8 + col str
      row = binarySpacePos 'B' 7 . take 7
      col = binarySpacePos 'R' 3 . drop 7
      findMissing (a : b : rest)
        | a + 1 == b = findMissing (b : rest)
        | otherwise = a + 1

  seats <- sort . map seatID . lines <$> readFile "input/day5"

  putStrLn $ "Part 1: " ++ show (last seats)
  putStrLn $ "Part 2: " ++ show (findMissing seats)

binarySpacePos :: Char -> Int -> String -> Int
binarySpacePos upperChar len = fst . foldl folder (0, 0)
  where
    folder (pos, index) char
      | char == upperChar = (pos + 2 ^ (len - index - 1), index + 1)
      | otherwise = (pos, index + 1)
